File: bookstore/mern/controllers/author.controller.js
import { authorService } from '../services/index.js'
import { deleteBooks, errorHandler } from '../shared/helpers.js'

export default class AuthorController {
  async get(req, res) {
    const authors = await authorService.get()
    return res.json(authors)
  }

  async create(req, res, next) {
    return errorHandler(async () => {
      const author = await authorService.create(req.body.name)
      return res.json(author)
    }, next)
  }

  async change(req, res, next) {
    return errorHandler(async () => {
      const author = await authorService.change(req.body._id, req.body.name)
      return res.json(author)
    }, next)
  }

  async delete(req, res, next) {
    return errorHandler(async () => {
      const id = await authorService.delete(req.body._id)
      await deleteBooks({ _author: id })
      return res.json(id)
    }, next)
  }
}

File: bookstore/spring-backend/src/main/kotlin/masha/bookstore/controllers/AuthorController.kt
package masha.bookstore.controllers

import masha.bookstore.dtos.AuthorDto
import masha.bookstore.interfaces.IAuthorService
import masha.bookstore.interfaces.IBookService
import masha.bookstore.models.Author
import masha.bookstore.shared.Auth
import masha.bookstore.shared.Role
import masha.bookstore.shared.Route
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(Route.AUTHOR)
@CrossOrigin(originPatterns = ["*"])
class AuthorController(
    private val authorService: IAuthorService,
    private val bookService: IBookService
) {

    @GetMapping
    fun getAll(): List<AuthorDto> = authorService.getAll().map { it.toDto() }

    @PostMapping
    fun create(
        @RequestBody dto: AuthorDto,
        @RequestHeader("Authorization") authorization: String?
    ): AuthorDto {
        Auth.verify(authorization, Role.ONLY_ADMIN)
        return authorService.create(dto.name).toDto()
    }

    @PatchMapping
    fun change(
        @RequestBody dto: AuthorDto,
        @RequestHeader("Authorization") authorization: String?
    ): AuthorDto {
        Auth.verify(authorization, Role.ONLY_ADMIN)
        return authorService.change(dto.id, dto.name).toDto()
    }

    @DeleteMapping
    fun delete(
        @RequestBody dto: AuthorDto,
        @RequestHeader("Authorization") authorization: String?
    ): Int {
        Auth.verify(authorization, Role.ONLY_ADMIN)
        val author: Author = authorService.getOne(dto.id)
        author.books.forEach { bookService.delete(it.id) }
        return authorService.delete(dto.id)
    }
}